python/lsst/ts/ATHexapod/controller.py

```diff
@@ -117,7 +128,7 @@ def is_connected(self):
         return response
```

Comment:
I *strongly* suggest that you make the # of lines of reply an argument (instead of bool `multi_line`) and just read that many times. It would make the code *much* simpler and avoid the danger of silently returning, say, 3 lines when 5 was expected. The new code would look like this and avoid a pointless read-one-char-at-a-time tight Python loop (which is *very* inefficient).

Taking both suggestions into account leaves the end of the method looking like this (where `[:-1]` or the equivalent `[0:-1]` strips the trailing `\n`):
```
try:
    response = b""
    for i in range(n_lines):
        response += await asyncio.wait_for(self.reader.readuntil("\n"), timeout=2)
except asyncio.TimeoutError:
    self.log.warning("Timed out waiting for response from controller. Result is incomplete")

return response.decode()[:-1]
```

Reply:
Finally, please consider allowing the code to raise TimeoutError instead of swallowing it. The problem with your current code is that everyone who calls `write_command` has to deal with return data that may be corrupted by being truncated. That complicates a *lot* of methods.

If you allow the exception to be raised then you can probably handle the exception in some much higher level (whatever code is calling the various methods that call write_command, not all those methods).

Even if you have to handle the error in each method, at least you have a clue there was a problem! As it stands you have to test the return data (and if you forget then god knows what strange symptoms occur).

That leaves you with:
```
try:
    response = b""
    for i in range(n_lines):
        response += await asyncio.wait_for(self.reader.readuntil("\n"), timeout=2)
except asyncio.TimeoutError:
    self.log.warning(f"Read {i+1} lines but expected {n_lines} in response to command {cmd}")
    raise
```

Reply:
According to Tiago the controller returns data in random chunks which is why he reads character by character.

Reply:
@tribeiro I'm puzzled by Eric's comment above. The point of `readuntil` is to eliminate the need to worry about the controller sending data in chunks. It waits until all the data that we want arrives. Did you try `readuntil`?

Reply:
Perhaps try the code I suggested?

Note that specifying the number of lines as an argument is orthogonal to the issue of chunking and reading char by char. So you could do that even if you find you have to read char by char (though I really don't see how it is possible that readuntil would fail).